network/p2p/tracer/internal/cache.go

```diff
@@ -0,0 +1,84 @@
+package internal
+
+import (
+	"github.com/rs/zerolog"
+
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/module"
+	herocache "github.com/onflow/flow-go/module/mempool/herocache/backdata"
+	"github.com/onflow/flow-go/module/mempool/herocache/backdata/heropool"
+	"github.com/onflow/flow-go/module/mempool/stdmap"
+	p2pmsg "github.com/onflow/flow-go/network/p2p/message"
+)
+
+type RPCSentCacheConfig struct {
+	sizeLimit uint32
+	logger    zerolog.Logger
+	collector module.HeroCacheMetrics
+}
+
+// rpcSentCache cache that stores rpcSentEntity. These entity's represent RPC control messages sent from the local node.
+type rpcSentCache struct {
+	// c is the underlying cache.
+	c *stdmap.Backend
+}
+
+// newRPCSentCache creates a new *rpcSentCache.
+// Args:
+// - config: record cache config.
+// Returns:
+// - *rpcSentCache: the created cache.
+// Note that this cache is intended to track control messages sent by the local node,
+// it stores a RPCSendEntity using an Id which should uniquely identifies the message being tracked.
+func newRPCSentCache(config *RPCSentCacheConfig) (*rpcSentCache, error) {
+	backData := herocache.NewCache(config.sizeLimit,
+		herocache.DefaultOversizeFactor,
+		heropool.LRUEjection,
+		config.logger.With().Str("mempool", "gossipsub=rpc-control-messages-sent").Logger(),
+		config.collector)
+	return &rpcSentCache{
+		c: stdmap.NewBackend(stdmap.WithBackData(backData)),
+	}, nil
+}
+
+// init initializes the record cached for the given messageID if it does not exist.
+// Returns true if the record is initialized, false otherwise (i.e.: the record already exists).
+// Args:
+// - flow.Identifier: the messageID to store the rpc control message.
+// - p2p.ControlMessageType: the rpc control message type.
+// Returns:
+// - bool: true if the record is initialized, false otherwise (i.e.: the record already exists).
+// Note that if init is called multiple times for the same messageID, the record is initialized only once, and the
+// subsequent calls return false and do not change the record (i.e.: the record is not re-initialized).
+func (r *rpcSentCache) init(messageID flow.Identifier, controlMsgType p2pmsg.ControlMessageType) bool {
```

Comment:
I propose repositioning the `iHaveRPCSentEntityID` method to this file and consolidating it with the `newRPCSentEntity` function as follows:

```go
func rpcSentEntityID(topic string, messageId string, controlMsgType p2pmsg.ControlMessageType) flow.Identifier {
	return flow.MakeIDFromFingerPrint([]byte(fmt.Sprintf("%s%s%s", topic, messageId, controlMsgType)))
}
```

This modification promotes efficient computing by saving one hash computation per message. Currently, we're processing two hashes for each message, which seems superfluous.

In addition, I recommend a substantial refactor of the `init` method to improve its alignment with common software architecture principles such as cohesion, coupling, and simplicity:

1. The method name `init` should be changed to `add`. This aligns with the standard `Add` operations found in our mempools, which will improve understandability due to its familiar behavior. 

2. Its signature should be modified to `add(topic string, messageId string, controlMsgType p2pmsg.ControlMessageType)`. This change enhances cohesion and reduces confusion by clarifying the type of records it maintains. The current `init(messageID flow.Identifier, controlMsgType p2pmsg.ControlMessageType) bool` signature displays low cohesion and high coupling, requiring the caller to comprehend and compute the `messageID` externally. Given that the `messageID` is merely used as a key to keep records, this understanding should be internal to the cache. The current setup expects the caller to comprehend this internal aspect, which is not aligned with the best practices. Furthermore, the `messageID` as a `flow.Identifier` provides a generic representation, leaving ambiguity about its actual meaning. This issue can be resolved by utilizing the `rpcSentEntityID` method defined above to compute the `messageID`.

Such changes adhere to the principles of software architecture, namely encapsulation, cohesion, and low coupling. By hiding internal details like the computation of `messageID`, we achieve better encapsulation. The refactoring also improves cohesion because the new `add` method will perform a more focused, single responsibility: adding a record based on specific input parameters. Additionally, the coupling is reduced as the new method is less reliant on the external computation of `messageID`.

Reply:
https://github.com/onflow/flow-go/pull/4543/commits/60b764dc9ce99628578fb3407689316c3e22c3f9